src/main/java/jnr/enxio/channels/Common.java

```diff
@@ -44,13 +44,19 @@ int getFD() {
 
     int read(ByteBuffer dst) throws IOException {
 
-        ByteBuffer buffer = ByteBuffer.allocate(dst.remaining());
+        int n;
+        if (!dst.hasArray() && !dst.isDirect()) {
```

Comment:
Why do this additional check at all? Why not call `Native.read` directly with `dst`. I can't see why the new allocation is required even for non-direct mode. `Native.read` appears to be agnostic to the type of byte buffer, and if there is something to be done, then that's the JNI call-site and the best place for those concerns.